sql/distsql: TestAggregatorAgainstProcessor failed

[(sql/distsql).TestAggregatorAgainstProcessor failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2679761&tab=buildLog) on [master@68b86d57c7f36b9b6926d1fb82fda0cb5674a8f1](https://github.com/cockroachdb/cockroach/commits/68b86d57c7f36b9b6926d1fb82fda0cb5674a8f1):

```
        36: [NULL NULL NULL 1.0765524423785964 0.0]
        37: [NULL NULL NULL 1.7014117331926443e+38 0.5819884116693155]
        38: [NULL NULL NULL -0.4426748936213637 0.0]
        39: [NULL 1.0787054592952081e-09 -1.9225409009185915e-19 0.4035910905207397 0.0]
        40: [NULL NULL NULL 0.7403943210754997 0.0]
        41: [NULL -2.409743598806959e-10 NULL -0.11366294145824818 0.0]
        42: [NULL 4.989206757505815e-10 5.3409870495970017e-20 0.42870328532390284 0.0]
        43: [NULL 2.0554132743051876e-09 NULL -0.9138972003586194 0.0]
        44: [-1.6915796002905105e+23 NULL 9.604093447827515e-20 0.10521226056701288 0.796675157500248]
        45: [NULL NULL NULL NULL 0.0]
        46: [NULL NULL NULL -0.938464494739478 0.0]
        47: [NULL NULL -1.3444413642763673e-19 -1.4537330274919766 NaN]
        48: [NULL 3.708708067984454e-10 0.0009543222163959106 NULL 0.0]
        49: [NULL NULL NULL -0.1207986994081186 0.0]
        50: [NULL NULL NULL NULL 0.0]
        51: [NULL NULL NULL 0.014980862235155712 0.0]
        52: [NULL NULL -3.123037919156088e-20 0.9180222190423484 0.0]
        53: [NULL 1.6479035136421945e-09 -4.673109219287149e-19 NULL 0.0]
        54: [NULL -6.075267071684981e-10 NULL 0.878403429669872 0.0]
        55: [NULL NULL NULL -1.349290817730205 0.0]
        56: [NULL 6.842654173325237e-10 NULL -0.1539850006341148 0.0]
        57: [NULL NULL 2.9136345431316945e-19 0.634650222717432 0.0]
        58: [NULL 5.862098053598851e-10 NULL NULL 0.0]
        59: [NULL NULL NULL NULL 0.0]
        60: [NULL NULL NULL 1.0 NaN]
        61: [NULL 4.879000023022722e-10 -5.063177333566852e-19 0.42448487303052174 0.0]
        62: [NULL NULL 2.5806363467405467e-19 -1.0 0.0]
        63: [NULL 2.710139759651961e-09 6.452431106841037e-20 NULL 0.0]
        64: [5.970445964467199e+22 NULL NaN -0.11776685122039132 0.4881379048688079]
        65: [NULL 1.2391341050553218e-05 NULL -0.028043448867100418 0.0]
        66: [NULL 2.5680745945435276e-10 NULL 1.7296108960662466 0.0]
        67: [NULL NULL NULL 0.9860813506204302 0.0]
        68: [NULL 9.680336456501586e-10 3.0694420101858523e-19 -1.1351803712841124 0.0]
        69: [NULL NULL 8.640882707672539e-19 -1.8139679617123 0.0]
        70: [NULL NULL NULL -2.0295401896097505 0.0]
        71: [NULL NULL NULL -1.670459857686138 0.0]
        72: [NULL 1.0456143599592293e-10 6.193184365922846e-19 -0.22010251690923244 0.0]
        73: [NULL NULL -1.2516267352273838e-19 NULL 0.0]
        74: [NULL +Inf NULL 0.2847239349743496 NaN]
        75: [NULL NULL -1.2936049305085159e-18 -1.7168413567507044 0.0]
        76: [NULL NULL NaN 0.5464140668211925 NULL]
        77: [NULL NULL -5.820359218024855e-19 3.4028234663852886e+38 NaN]
        78: [NULL -8.001736714561304e-09 NULL NULL 0.0]
        79: [NULL 2.8255914200114123e-09 NULL -0.5497589295485126 0.0]
        80: [NULL NaN NULL -2.142483136846359 0.0]
        81: [NULL -1.0523958131758973e-07 NULL NULL 0.0]
        82: [NULL -3.3750764384812335e-10 -2.798184078082056e-20 0.9772981283716842 0.0]
        83: [NULL NULL 7.478488303054997e-19 -1.170857343709085 0.0]
        84: [NULL NULL -4.764783217018356e-19 -1.1350108106765164 0.0]
--- FAIL: TestAggregatorAgainstProcessor (5.92s)

```

<details><summary>More</summary><p>
Parameters:

- TAGS=
- GOFLAGS=-parallel=4

```
make stressrace TESTS=TestAggregatorAgainstProcessor PKG=./pkg/sql/distsql TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```


[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2ATestAggregatorAgainstProcessor.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

--- a/pkg/sql/distsql/columnar_utils_test.go
+++ b/pkg/sql/distsql/columnar_utils_test.go
@@ -12,10 +12,11 @@ package distsql
 
 import (
 	"context"
-	"fmt"
 	"math"
 	"math/rand"
+	"sort"
 	"strconv"
+	"strings"
 	"testing"
 
 	"github.com/cockroachdb/cockroach/pkg/base"
@@ -257,14 +258,6 @@ func verifyColOperator(t *testing.T, args verifyColOperatorArgs) error {
 			procRows, colOpRows, procMetas, colOpMetas)
 	}
 
-	printRowsOutput := func(rows [][]string) string {
-		res := ""
-		for i, row := range rows {
-			res = fmt.Sprintf("%s\n%d: %v", res, i, row)
-		}
-		return res
-	}
-
 	datumsMatch := func(expected, actual string, typ *types.T) (bool, error) {
 		switch typ.Family() {
 		case types.FloatFamily:
@@ -303,54 +296,38 @@ func verifyColOperator(t *testing.T, args verifyColOperatorArgs) error {
 			colIdxsToCheckForEquality[i] = i
 		}
 	}
+
 	if args.anyOrder {
-		used := make([]bool, len(colOpRows))
-		for i, expStrRow := range procRows {
-			rowMatched := false
-			for j, retStrRow := range colOpRows {
-				if used[j] {
-					continue
-				}
-				foundDifference := false
+		// The rows are allowed to be in any order, so in order to use the
+		// ordered comparison below we will sort rows from both the processor
+		// and the operator lexicographically.
+		getLessFn := func(rows [][]string) func(int, int) bool {
+			return func(i, j int) bool {
 				for _, colIdx := range colIdxsToCheckForEquality {
-					match, err := datumsMatch(expStrRow[colIdx], retStrRow[colIdx], args.pspec.ResultTypes[colIdx])
-					if err != nil {
-						return errors.Errorf("error while parsing datum in rows\n%v\n%v\n%s",
-							expStrRow, retStrRow, err.Error())
-					}
-					if !match {
-						foundDifference = true
-						break
+					if cmp := strings.Compare(rows[i][colIdx], rows[j][colIdx]); cmp != 0 {
+						return cmp < 0
 					}
 				}
-				if !foundDifference {
-					rowMatched = true
-					used[j] = true
-					break
-				}
-			}
-			if !rowMatched {
-				return errors.Errorf("different results: no match found for row %d of processor output\n"+
-					"processor output:%s\n\ncolumnar operator output:%s",
-					i, printRowsOutput(procRows), printRowsOutput(colOpRows))
+				return false
 			}
 		}
-	} else {
-		for i, expStrRow := range procRows {
-			retStrRow := colOpRows[i]
-			// anyOrder is false, so the result rows must match in the same order.
-			for _, colIdx := range colIdxsToCheckForEquality {
-				match, err := datumsMatch(expStrRow[colIdx], retStrRow[colIdx], args.pspec.ResultTypes[colIdx])
-				if err != nil {
-					return errors.Errorf("error while parsing datum in rows\n%v\n%v\n%s",
-						expStrRow, retStrRow, err.Error())
-				}
-				if !match {
-					return errors.Errorf(
-						"different results on row %d;\nexpected:\n%s\ngot:\n%s",
-						i, expStrRow, retStrRow,
-					)
-				}
+		sort.Slice(procRows, getLessFn(procRows))
+		sort.Slice(colOpRows, getLessFn(colOpRows))
+	}
+
+	for i, expStrRow := range procRows {
+		retStrRow := colOpRows[i]
+		for _, colIdx := range colIdxsToCheckForEquality {
+			match, err := datumsMatch(expStrRow[colIdx], retStrRow[colIdx], args.pspec.ResultTypes[colIdx])
+			if err != nil {
+				return errors.Errorf("error while parsing datum in rows\n%v\n%v\n%s",
+					expStrRow, retStrRow, err.Error())
+			}
+			if !match {
+				return errors.Errorf(
+					"different results on row %d;\nexpected:\n%s\ngot:\n%s",
+					i, expStrRow, retStrRow,
+				)
 			}
 		}
 	}

